docs/howtos/when-to-use-what-in-the-ffi.md

```diff
@@ -0,0 +1,133 @@
+   Additionally, for types T such that `&T: Sync + Send`, or that you
+   need to call `&mut self` method, this is the safest choice.
+
+   Additionally, this will ensure panic-safety, as you'll poison your Mutex.
+
+   The [`ffi_support::handle_map` docs](https://docs.rs/ffi-support/*/ffi_support/handle_map/index.html)
+   are good, and under `ConcurrentHandleMap` include an example of how to set
+   this up. You can also look at most of the FFI crates, as they do this (with
+   the exception of `rc_log`, which has unique requirements).
+
+2. Using an opaque pointer. This is generally only recommended for rare cases
+   like the `PlacesInterruptHandle` (or the `LogAdapterState` from `rc_log`,
+   although it will probably eventually use a handle).
+
+   It's good if your synchronization or threading requirements are somewhat
+   complex and handled separately, such that the additional overhead of
+   the `ConcurrentHandleMap` is undesirable. You should probably talk to us
+   before adding another type that works this way, to make sure it's sound.
+
+   The [`ffi_support` docs](https://docs.rs/ffi-support/*/ffi_support/macro.implement_into_ffi_by_pointer.html)
+   discuss how to do this, or take a look at how it's done for
+   `PlacesInterruptHandle`).
+
+## Plain Old Data
+
+This includes both primitive values, strings, arrays, or arbitrarially nested
+structures containing them.
+
+### Primitives
+
+Specifically numeric primitives. these we'll tackle first since they're the
+easiest.
+
+In general, you can just pass them as you wish. There are a couple of
+exceptions/caveats. All of them are caused by JNA/Android issues (Swift has very
+good support for calling over the FFI), but it's our lowest common denominator.
+
+1. `bool`: Don't use it. JNA doesn't handle it well. Instead, use a numeric type
+    (like `u8`) and represent 0 for false and 1 for true for interchange over the
+    FFI, converting back to a Kotlin `Boolean` or swift `Bool` after (as to
+    not expose this somewhat annoying limitation in our public API).
+
+2. `usize`/`isize`: These cause the structure size to be different based on the
+   platform. JNA does handle this if you use `NativeSize`, but it's awkward,
+   incompatible with it's Direct Mapping optimization (which we don't use but
+   want to in the future), and has more overhead than just using `i64`/`i32` for
+   `Long`/`Int`. (You can also use `u64`/`u32` for `Long`/`Int`, if you're certain the
+   value is not negative)
+
+3. `char`: I really don't see a reason you need to pass a single codepoint over the
+   FFI, but if someone needs to do this, they instead should just pass it as a `u32`.
+
+    If you do this, you should probably be aware of the fact that Java chars are 16
+    bit, and Swift `Character`s are actually strings (they represent Extended
+    Grapheme Clusters, not codepoints).
+
+### Strings
+
+These we pass as nul-terminated UTF-8 C-strings.
```

Comment:
nit: null?

Reply:
NUL is the ascii character, see https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_with_nul, `man ascii`, etc.

NULL is sometimes used but also means other things.

Reply:
Ah, Google auto-corrects nul-terminated to null, per https://en.wikipedia.org/wiki/Null-terminated_string

Definitely a nit. 😀